### PR TITLE
Graph

### PR DESCRIPTION
**Pull request by lxcyqx**
Complete user ranking by dynamically building graph while using Dijkstra's algorithm to find the shortest path to target user

### CODE REVIEW

File: capstone/src/main/java/database/DatabaseRetriever.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
+package database;
```
Comment (nicorod-google) [inline capstone/src/main/java/database/DatabaseRetriever.java]: Why did you move this from java/com/google/sps/search/ to java/database? We should keep everything under the java/com/google structure

File: capstone/src/main/java/com/google/sps/search/servlets/SearchResultsServlet.java [hunk 2]
```diff
@@ -26,21 +35,32 @@
 @WebServlet("/search-results")
 public class SearchResultsServlet extends AuthenticatedServlet {
 
+  private static double COMPLETE_MATCH = 10;
```
Comment (nicorod-google) [inline capstone/src/main/java/com/google/sps/search/servlets/SearchResultsServlet.java]: nit: make this final

File: capstone/src/main/java/database/DatabaseRetriever.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.sps.graph.UserVertex;
+
+public class DatabaseRetriever implements Serializable {
+
+  /** */
+  private static final long serialVersionUID = 1L;
```
Comment (nicorod-google) [inline capstone/src/main/java/database/DatabaseRetriever.java]: I'm not sure this is used anywhere? If it's necessary, then please fill out the comment because I don't know what it does as well as make it a constant since it's final.

**Review by nicorod-google — CHANGES_REQUESTED** [review]

File: capstone/src/main/java/com/google/sps/graph/VertexComparator.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+public class VertexComparator<E> implements Comparator<Vertex<E>> {
+
+  @Override
+  public int compare(Vertex<E> v1, Vertex<E> v2) {
+    if (v1.getDistance() < v2.getDistance()) {
+      return -1;
+    } else if (v1.getDistance() > v2.getDistance()) {
+      return 1;
+    }
```
Comment (g-asweet) [inline capstone/src/main/java/com/google/sps/graph/VertexComparator.java]:
```java
return Double.compare(v1.getDistance(), v2.getDistance());
```

File: capstone/src/main/java/com/google/sps/graph/Dijkstra.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+          // remember path to this vertex
+          adjacent.setPrev(e);
+          // Trigger reheapify since vertex has been mutated
+          if (minHeap.contains(adjacent)) {
+            minHeap.add(minHeap.remove());
+          } else {
+            minHeap.add(adjacent);
+          }
+        }
+      }
```
Comment (g-asweet) [inline capstone/src/main/java/com/google/sps/graph/Dijkstra.java]: I think you should move 

```java
for (E e : outgoingEdges) {
  // Get the vertex on the other side of the edge
  V adjacent = e.getDestVertex();
  if (!visitedSet.contains(adjacent)) {
     ...
  }
}
```

to wrap all of this, as you never want to visit the same point twice.

**Review by g-asweet — CHANGES_REQUESTED** [review]

File: capstone/src/main/java/com/google/sps/graph/Dijkstra.java (hunk 5 above)
Comment (g-asweet) [inline capstone/src/main/java/com/google/sps/graph/Dijkstra.java]: I'd pull this out into a `processNeighbors(current, visitedSet, minHeap)` method

Comment (g-asweet) [reply capstone/src/main/java/com/google/sps/graph/Dijkstra.java]: Unless you want to share a link to whichever resource you found online which most helped in shaping your dijkstra implementation, which I coudl be wrong about

Comment (g-asweet) [reply capstone/src/main/java/com/google/sps/graph/Dijkstra.java]: https://www.vogella.com/tutorials/JavaAlgorithmsDijkstra/article.html Hmm this seems to be pretty close to what you have

File: capstone/src/main/java/com/google/sps/search/servlets/SearchResultsServlet.java [hunk 6]
```diff
@@ -50,13 +70,55 @@ public void doGet(String userId, HttpServletRequest request, HttpServletResponse
       for (Entity result : pq.asIterable()) {
         try {
           users.add(User.fromEntity(result));
+          namesScore = this.runDijkstra(namesScore, userId, searchString, User.fromEntity(result));
```
Comment (g-asweet) [inline capstone/src/main/java/com/google/sps/search/servlets/SearchResultsServlet.java:72]: HIGH level, I'm concerned that every time somebody types in a new character, you're getting a list of names using your trie implementation AND following up that entire result list with a call to dijkstra for every user in the result list......

If you're going to go a pathfinding route, I would suggest precalculating and saving distances to the database, or at least caching results.

File: capstone/src/main/java/database/DatabaseRetriever.java (hunk 3 above)
Comment (g-asweet) [reply capstone/src/main/java/database/DatabaseRetriever.java]: I believe this has something to do with being a serializable object.

https://www.baeldung.com/java-serial-version-uid#serial-version-uid

If you're saving data into some place, and then updating your logic, and trying to load that serialized data back, you could end up in a mess if they objects weren't the same. This versioning allows you to identify those changes.

**Review by g-asweet — DISMISSED** [review]
Last comment is that this is a lot of change for the second to last day of your internship. I'd take a crack at making it work today, but I wouldn't stress over making sure you've submitted this. Documenting your existing features and future plans of solving a graph problem will do you just as well as finishing this.

File: capstone/src/main/java/com/google/sps/search/servlets/SearchResultsServlet.java (hunk 6 above)
Comment (lxcyqx) [reply capstone/src/main/java/com/google/sps/search/servlets/SearchResultsServlet.java:72]: This servlet is only called when the user presses enter to list the user results, so it'll only do the pathfinding once when the user presses enter and not at each new character. Caching would still be useful though

File: capstone/src/main/java/database/DatabaseRetriever.java (hunk 1 above)
Comment (lxcyqx) [reply capstone/src/main/java/database/DatabaseRetriever.java]: I just created a new package and it defaulted to that. I've now changed it to be under the java/com/google structure. Is there a reason for it to be under that though?

Comment (nicorod-google) [reply capstone/src/main/java/database/DatabaseRetriever.java]: That's the structure for java code at Google: 
java/com/google/a/b/c/ClassName.java if it will be part of any production system

**Review by nicorod-google — APPROVED** [review]
